src/gateway/WebSocketClient.ts

```diff
@@ -0,0 +1,128 @@
+import WebSocket from 'ws';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { EVENTS, Payload } from './constants';
+import { GATEWAY_OPCODES, GATEWAY_CLOSE_EVENT_CODES } from '../util/Constants';
+const IDENTIFY_TIMEOUT = 4000;
+
+export default class WebSocketClient {
+  private _heartBeatInterval!: NodeJS.Timeout;
+  private _intents: number;
+  private _lastHeartbeatAck = true;
+  private _seq!: number;
+  private _sessionID!: string;
+  private _token: string;
+  private _url: string;
+  private _ws!: WebSocket;
+
+  constructor(token: string, wsURL: string, intents: number) {
+    this._token = token;
+    this._url = wsURL;
+    this._intents = intents;
+
+    this._identify = this._identify.bind(this);
+    this._onClose = this._onClose.bind(this);
+    this._onMessage = this._onMessage.bind(this);
+  }
+
+  connect() {
+    this._ws = new WebSocket(this._url);
+    this._ws.on('open', this._identify);
+    this._ws.on('close', this._onClose);
+    this._ws.on('message', this._onMessage);
+  }
+
+  send(op: GATEWAY_OPCODES, data: any) {
+    if (this._ws.readyState === WebSocket.OPEN) {
+      this._ws.send(JSON.stringify({
+        op,
+        d: data,
+      }));
+    }
+  }
+
```

Comment:
Should we add some indication to show whether the payload was actually sent or not? Probably via a bool return type

Reply:
I think we can add that to the todos/.
Later all send/ will need to be queued. To handle stuff like identify rl or delay when reconnecting.
Waiting for your approval for the whole PR. If Deadshot wants to implement a boolean return though that's something we will use in the future